src/Features/CSharp/Portable/CodeFixes/AddParameterXmlDocumentation/AddParameterXmlDocumentationCodeFixProvider.cs

```diff
@@ -0,0 +1,316 @@
+                        adjacentAttributeIsBefore, isTypeParameter: true),
+                    diagnostic);
+            }
+        }
+
+        /// <summary>
+        /// Finds a xml documentation tag with a name attribute at which another a new tag can be inserted.
+        /// </summary>
+        /// <param name="parameterName">The parameter of the new tag</param>
+        /// <param name="allParameterNames">All the parameter of which <paramref name="parameterName"/> is one.</param>
+        /// <param name="tokenWithDocumentationComment"></param>
+        /// <param name="elementKind">Specifies which elements can be found.</param>
+        /// <param name="adjacentNameAttribute">The name attribute of the adjacent parameter. If null no parameter was found.</param>
+        /// <param name="adjacentAttributeIsBefore">Indicates if the found adjacent attribute is before or after the parameter.</param>
+        private static void FindAdjacentParameterDocumentation(string parameterName, IEnumerable<string> allParameterNames, SyntaxToken tokenWithDocumentationComment,
+            XmlNameAttributeElementKind elementKind, out XmlNameAttributeSyntax adjacentNameAttribute, out bool adjacentAttributeIsBefore)
+        {
+            FindAdjacentParameterDocumentationWalker.FindAdjacentParameterDocumenteden(parameterName, allParameterNames, tokenWithDocumentationComment,
+                elementKind, out adjacentNameAttribute, out adjacentAttributeIsBefore);
+        }
+
+        private class FindAdjacentParameterDocumentationWalker : CSharpSyntaxWalker
+        {
+            private FindAdjacentParameterDocumentationWalker(int parameterIndex, Dictionary<string, int> parameterIndices, XmlNameAttributeElementKind elementKind)
+                : base(SyntaxWalkerDepth.StructuredTrivia)
+            {
+                _parameterIndex = parameterIndex;
+                _parameterIndices = parameterIndices;
+                _elementKind = elementKind;
+            }
+
+            private int _parameterIndex = -1;
+            private Dictionary<string, int> _parameterIndices = new Dictionary<string, int>();
+
+            private XmlNameAttributeSyntax _documentedParameterBefore = null;
+            private int _documentedParameterBeforeIndex = -1;
+            private XmlNameAttributeSyntax _documentedParameterAfter = null;
+            private int _documentedParameterAfterIndex = int.MaxValue;
+
+            private XmlNameAttributeElementKind _elementKind;
+
+            public static void FindAdjacentParameterDocumenteden(string parameterName, IEnumerable<string> allParameterNames,
+                SyntaxToken tokenWithDocumentationComment, XmlNameAttributeElementKind elementKind, out XmlNameAttributeSyntax adjacentNameAttribute, out bool adjacentAttributeIsBefore)
+            {
+                var parameterIndices = new Dictionary<string, int>();
+                int index = 0;
+                foreach (var param in allParameterNames)
```

Comment:
It's a *little* smelly that you foreach over an `IEnumerable<...>` in a way that looks like you really want a for loop. Perhaps you should just pass the `BaseParameterListSyntax` here so that you can loop over the parameters a bit more naturally?